core/src/main/java/org/apache/iceberg/GenericManifestFile.java

```diff
@@ -84,15 +84,24 @@ public GenericManifestFile(Schema avroSchema) {
     }
   }
 
+  /**
+   * @deprecated since 1.5.0, will be removed in 1.6.0. Use {@link
+   *     GenericManifestFile#GenericManifestFile(InputFile, int, Long)} with a snapshot-id
+   */
+  @Deprecated
   GenericManifestFile(InputFile file, int specId) {
+    this(file, specId, null);
+  }
+
+  GenericManifestFile(InputFile file, int specId, Long snapshotId) {
```

Comment:
I think I understand the rationale for adding this, but I'm pretty sure that we didn't do it this way on purpose. By adding this option, you can easily pass in the snapshot ID for a manfiest file so that you're deciding when it is created what snapshot it has.

This is something that we wanted to avoid so that people didn't accidentally corrupt metadata by setting the snapshot ID directly. Instead, callers can only construct manifests with snapshot IDs that will be inherited from manifest list metadata. At least, that's the idea. It's probably possible to work around it, but I think it is more correct for people to not have the option when creating the manifest metadata.

Also, I think this is something we use so that you can write a manifest of files (like how Flink stores checkpoint state) and later add it to the table once the snapshot ID is known.

Reply:
This change was required otherwise we'll end up writing a null for the `added_snapshot_id`. 

This method is only used in cases for V1. Mostly in the test folder (as the PR shows), but also when there is no manifest-list, and it directly points to the manifest:

![image](https://github.com/apache/iceberg/assets/1134248/a5cb7aec-6536-43e7-9b5c-57b4e478b8db)

I've reverted this change, and used the `copyOf` instead. LMKWYT

Reply:
`copyOf` loses the reference to the file, and therefore the length will be `null`, resulting in an NPE. Let me dig into this.